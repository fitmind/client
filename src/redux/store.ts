import { Store, createStore, applyMiddleware } from 'redux';
import { ApplicationState, storeInitialState, createRootReducer } from './reducers/root.reducer';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import { rootSaga } from './saga/root.saga';
import { History, createBrowserHistory } from 'history';
import { routerMiddleware } from 'connected-react-router';

// We use hash history because this example is going to be hosted statically.
// Normally you would use browser history.
export const history: History = createBrowserHistory();

export const configureStore = (initialState: ApplicationState): Store<ApplicationState> => {
    const sagaMiddleware = createSagaMiddleware();
    const composeEnhancers = composeWithDevTools({});
    let store = createStore(
        createRootReducer(history),
        initialState,
        composeEnhancers(applyMiddleware(routerMiddleware(history), sagaMiddleware)),
    );

    sagaMiddleware.run(rootSaga);
    return store;
};

const store: Store<ApplicationState> = configureStore(storeInitialState);

export default store;
