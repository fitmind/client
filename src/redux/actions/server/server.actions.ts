import { Action } from 'redux';
import { CustomerUserResponse } from '../../../interfaces/responses/customer-user-response';

export const ServerActions = {
    LOGIN_CUSTOMER_USER: 'LOGIN_CUSTOMER_USER',
    SET_CUSTOMER_USER: 'SET_CUSTOMER_USER',
    SIGNUP_CUSTOMER_USER: 'SIGNUP_CUSTOMER_USER',
    LOGOUT_USER: 'LOGOUT_USER',
    LOGOUT_USER_SUCCESS: 'LOGOUT_USER_SUCCESS',
};

export interface ServerActionTypes {
    type: string;
    errorMessage?: string;
    customerUser?: CustomerUserResponse;
}

export interface UserLoginActionInterface extends Action<'LOGIN_CUSTOMER_USER'> {
    email: string;
    password: string;
}

export const userLoginAction = ({ email, password }): UserLoginActionInterface => ({
    type: 'LOGIN_CUSTOMER_USER',
    email,
    password,
});

export interface setCustomerUserActionInterface extends Action<'SET_CUSTOMER_USER'> {
    customerUser: CustomerUserResponse;
}

export const setCustomerUserAction = (customerUser: CustomerUserResponse): setCustomerUserActionInterface => ({
    type: 'SET_CUSTOMER_USER',
    customerUser,
});

export interface CustomerSignUpActionInterface extends Action<'SIGNUP_CUSTOMER_USER'> {
    email: string;
    firstName: string;
    lastName: string;
    password: string;
    interestedInExperiseAreas: string[];
    description: string;
    phone: string;
}

export const customerSignUpAction = ({
    email,
    firstName,
    lastName,
    password,
    interestedInExperiseAreas,
    description,
    phone,
}: {
    email: string;
    firstName: string;
    lastName: string;
    password: string;
    interestedInExperiseAreas: string[];
    description: string;
    phone: string;
}): CustomerSignUpActionInterface => ({
    type: 'SIGNUP_CUSTOMER_USER',
    email,
    firstName,
    lastName,
    password,
    interestedInExperiseAreas,
    description,
    phone,
});
export type UserLogoutActionInterface = Action<'LOGOUT_USER'>;

export const userLogoutAction = (): UserLogoutActionInterface => ({
    type: 'LOGOUT_USER',
});

export type UserLogoutSuccessActionInterface = Action<'LOGOUT_USER_SUCCESS'>;

export const userLogoutSuccessAction = (): UserLogoutSuccessActionInterface => ({
    type: 'LOGOUT_USER_SUCCESS',
});
