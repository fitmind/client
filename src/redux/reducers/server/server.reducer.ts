import { Reducer } from 'redux';
import { serverActions, ServerActionTypes } from '../../actions/server.actions';
import { CustomerUser } from '../../../interfaces/customer-user';
import { ExpertUser } from '../../../interfaces/expert-user';

export interface ServerStateInterface {
    expertUser?: ExpertUser;
    customerUser?: CustomerUser;
}

export const initialCustomerUser: CustomerUser = {
    id: '',
    name: '',
    description: '',
    email: '',
    interestedInExpertiseAreas: [],
    pastBookings: [],
    futureBookings: [],
    pictureUrl: '',
};

export const initialExpert: ExpertUser = {
    id: '',
    name: '',
    description: '',
    email: '',
    isAnExpertIn: [],
    pastBookings: [],
    futureBookings: [],
    pictureUrl: '',
    weeklyAvailability: {
        monday: [],
        tuesday: [],
        wednesday: [],
        thursday: [],
        friday: [],
        saturday: [],
        sunday: [],
    },
    approvedStatus: '',
};

export const ServerInitialState: ServerStateInterface = {
    expertUser: initialExpert,
    customerUser: initialCustomerUser,
};

export const ServerReducer: Reducer<ServerStateInterface> = (
    state: ServerStateInterface = ServerInitialState,
    action: ServerActionTypes,
) => {
    switch (action.type) {
        case serverActions.SET_CUSTOMER_USER:
            return { ...state, customerUser: action.customerUser };
        case serverActions.DELETE_CUSTOMER_USER:
            return { ...state, customerUser: initialCustomerUser };
        case serverActions.SET_EXPERT_USER:
            return { ...state, expertUser: action.expertUser };
        case serverActions.DELETE_EXPERT_USER:
            return { ...state, expertUser: initialExpert };
        default:
            return state;
    }
};
