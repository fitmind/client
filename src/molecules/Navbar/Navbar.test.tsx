import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import NavBar from './Navbar';
import { Router } from 'react-router';
import store, { history, configureStore } from '../../redux/store';
import { ServerInitialState } from '../../redux/reducers/server-reducer/server.reducer';
import { Provider } from 'react-redux';
import { storeInitialState } from '../../redux/reducers/root.reducer';
import { merge, mergeDeepRight } from 'ramda';

describe('NavBar', () => {
    it('renders without crashing', () => {
        shallow(
            <Router history={history}>
                <NavBar />
            </Router>,
        );
    });

    describe('should match the snapshots', () => {
        it('should match when no one is logged in', () => {
            const component = renderer.create(
                <Provider store={store}>
                    <Router history={history}>
                        <NavBar />
                    </Router>
                </Provider>,
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('should match when the expert is logged in', () => {
            const mergedProps = mergeDeepRight(storeInitialState, {
                server: { customerUser: {}, expert: { _id: '123' } },
            });
            const storeWithExpert = configureStore(mergedProps);
            const component = renderer.create(
                <Provider store={storeWithExpert}>
                    <Router history={history}>
                        <NavBar />
                    </Router>
                </Provider>,
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('should match when the customer is logged in', () => {
            const mergedProps = mergeDeepRight(storeInitialState, {
                server: { customerUser: { _id: '123' }, expert: {} },
            });
            const storeWithExpert = configureStore(mergedProps);
            const component = renderer.create(
                <Provider store={storeWithExpert}>
                    <Router history={history}>
                        <NavBar />
                    </Router>
                </Provider>,
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    // needs snapshot for when the expert is logged in, the normal customer is logged in
});
