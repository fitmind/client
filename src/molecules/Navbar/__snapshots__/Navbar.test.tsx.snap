// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should match the snapshots should match when no one is logged in 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <span
    className="navbar-brand"
    onClick={[Function]}
  >
    Fitmind
  </span>
  <button
    aria-controls="basic-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="basic-navbar-nav"
  >
    <div
      className="mr-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Listings
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        About
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Contact
      </a>
    </div>
    <div
      className="mr-2 btn-group"
      role="group"
    >
      <button
        className="btn btn-outline-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Customers
      </button>
    </div>
    <button
      className="btn btn-outline-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Experts
    </button>
  </div>
</nav>
`;

exports[`NavBar should match the snapshots should match when the customer is logged in 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <span
    className="navbar-brand"
    onClick={[Function]}
  >
    Fitmind
  </span>
  <button
    aria-controls="basic-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="basic-navbar-nav"
  >
    <div
      className="mr-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Listings
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        About
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Contact
      </a>
    </div>
    <div
      className="mr-2 btn-group"
      role="group"
    >
      <button
        className="btn btn-outline-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Dashboard
      </button>
    </div>
    <button
      className="btn btn-outline-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </div>
</nav>
`;

exports[`NavBar should match the snapshots should match when the expert is logged in 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <span
    className="navbar-brand"
    onClick={[Function]}
  >
    Fitmind
  </span>
  <button
    aria-controls="basic-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="basic-navbar-nav"
  >
    <div
      className="mr-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Listings
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        About
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Contact
      </a>
    </div>
    <div
      className="mr-2 btn-group"
      role="group"
    >
      <button
        className="btn btn-outline-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Customers
      </button>
    </div>
    <button
      className="btn btn-outline-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Experts
    </button>
  </div>
</nav>
`;
